# EDA on Retail Sales Data

**Intern:** Majd Zouaoui
**Track:** Data Analytics (OIBSIP Level 1 — Task 1)
**Dataset:** Indian Retail Sales Dataset (Raw/Uncleaned) — Kaggle
**Date:** September 2026

##### This notebook performs Exploratory Data Analysis on a synthetic Indian retail sales dataset containing ~4,300 order records spanning 2020–2024. It covers data cleaning, descriptive statistics, time series trends, customer demographics, product performance, and business recommendations.

In [1]:
import kagglehub

# Download latest version
path = kagglehub.dataset_download("satyakidas07/retail-sales-dataset")

print("Path to dataset files:", path)

C:\Users\majdz\AppData\Local\Programs\Python\Python313\Lib\site-packages\tqdm\auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm


Path to dataset files: C:\Users\majdz\.cache\kagglehub\datasets\satyakidas07\retail-sales-dataset\versions\1


In [2]:
import pandas as pd 
df = pd.read_csv(r'C:\internship_oasis_online\OIBSIP\DataAnalytics-L1-EDARetailSales\data\retail_sales_dataset.csv')
df.head()

,order_id,order_date,customer_id,customer_name,age,gender,region,city,product_category,product_name,...,unit_price,discount_pct,sales_amount,profit,shipping_cost,payment_method,customer_satisfaction,return_flag,order_status,days_to_ship
0,ORD-03903,2024-08-15,CUST32603,Aarav Mehta,39.0,Other,Central,Raipur,Groceries,Sugar,...,346.32,NaN,4043.14,289.18,56.04,Debit Card,NaN,True,Returned,7.0
1,ORD-02198,2022-08-16,CUST07881,Sanjay Verma,28.0,Male,West,Mumbai,Beauty,Sunscreen,...,858.66,0.08,9147.31,5355.56,64.49,Debit Card,5.0,False,Delivered,2.0
2,ORD-01348,2021-08-29,CUST49296,Pooja Sharma,41.0,Other,West,Surat,Furniture,Sofa,...,25941.32,0.14,110156.97,36956.51,67.65,Debit Card,4.0,False,Delivered,5.0
3,ORD-02076,2022-07-05,CUST05587,Priya Reddy,29.0,Other,West,Ahmedabad,Furniture,Bed Frame,...,22071.92,0.17,262066.11,66644.99,54.47,EMI,3.0,False,Delivered,4.0
4,ORD-00287,2020-05-07,CUST58035,Ritu Sharma,9.0,Male,Central,Nagpur,Groceries,Cooking Oil,...,320.06,0.31,323.95,31.38,48.81,Net Banking,4.0,False,Pending,4.0


## Data Loading

The dataset was sourced from Kaggle ("Indian Retail Sales Dataset — Raw/Uncleaned") and stored locally in `data/`. I load it into a pandas DataFrame named `df`.

In [3]:
df.shape
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4310 entries, 0 to 4309
Data columns (total 21 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   order_id               4280 non-null   object 
 1   order_date             4280 non-null   object 
 2   customer_id            4280 non-null   object 
 3   customer_name          4280 non-null   object 
 4   age                    4150 non-null   float64
 5   gender                 4280 non-null   object 
 6   region                 4280 non-null   object 
 7   city                   4280 non-null   object 
 8   product_category       4280 non-null   object 
 9   product_name           4280 non-null   object 
 10  quantity               4170 non-null   float64
 11  unit_price             4280 non-null   float64
 12  discount_pct           4143 non-null   float64
 13  sales_amount           4280 non-null   float64
 14  profit                 4280 non-null   float64
 15  ship

In [4]:
df.shape

(4310, 21)

In [5]:
df.duplicated().sum()

np.int64(109)

In [6]:
for col in ['gender', 'region', 'product_category', 'payment_method', 'return_flag', 'order_status']:
    print(f"\n=== {col} ===")
    print(df[col].value_counts(dropna=False))


=== gender ===
gender
Other     1442
Male      1207
Female    1191
male        68
MALE        62
f           57
m           55
F           54
M           50
FEMALE      49
female      45
NaN         30
Name: count, dtype: int64

=== region ===
region
South      885
West       880
North      872
East       846
Central    797
NaN         30
Name: count, dtype: int64

=== product_category ===
product_category
Clothing       645
Groceries      627
Sports         625
Electronics    615
Beauty         612
Furniture      583
Books          573
NaN             30
Name: count, dtype: int64

=== payment_method ===
payment_method
UPI                 761
Credit Card         734
Cash on Delivery    717
Debit Card          700
EMI                 695
Net Banking         673
NaN                  30
Name: count, dtype: int64

=== return_flag ===
return_flag
False    3657
True      623
NaN        30
Name: count, dtype: int64

=== order_status ===
order_status
Delivered    2666
Returned      623
Cancel

In [7]:
df['order_date'].head(20)

0     2024-08-15
1     2022-08-16
2     2021-08-29
3     2022-07-05
4     2020-05-07
5     02/09/2024
6     2021-09-28
7     2023-06-10
8     2020-07-24
9     2020-02-09
10    2020-06-04
11    2022-01-28
12    2022-09-14
13    2024-01-21
14    2023-01-04
15    2024-03-10
16    2022-03-14
17    06/11/2023
18    2024-12-17
19    2020-01-19
Name: order_date, dtype: object

In [8]:
df[df['order_id'].isnull()].head()

,order_id,order_date,customer_id,customer_name,age,gender,region,city,product_category,product_name,...,unit_price,discount_pct,sales_amount,profit,shipping_cost,payment_method,customer_satisfaction,return_flag,order_status,days_to_ship
66,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
144,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
432,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
516,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
630,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN


In [9]:
df['age'].sort_values().head(10)
df['age'].sort_values().tail(10)

4052   NaN
4058   NaN
4063   NaN
4118   NaN
4120   NaN
4128   NaN
4133   NaN
4153   NaN
4269   NaN
4295   NaN
Name: age, dtype: float64

In [10]:
df.head()

,order_id,order_date,customer_id,customer_name,age,gender,region,city,product_category,product_name,...,unit_price,discount_pct,sales_amount,profit,shipping_cost,payment_method,customer_satisfaction,return_flag,order_status,days_to_ship
0,ORD-03903,2024-08-15,CUST32603,Aarav Mehta,39.0,Other,Central,Raipur,Groceries,Sugar,...,346.32,NaN,4043.14,289.18,56.04,Debit Card,NaN,True,Returned,7.0
1,ORD-02198,2022-08-16,CUST07881,Sanjay Verma,28.0,Male,West,Mumbai,Beauty,Sunscreen,...,858.66,0.08,9147.31,5355.56,64.49,Debit Card,5.0,False,Delivered,2.0
2,ORD-01348,2021-08-29,CUST49296,Pooja Sharma,41.0,Other,West,Surat,Furniture,Sofa,...,25941.32,0.14,110156.97,36956.51,67.65,Debit Card,4.0,False,Delivered,5.0
3,ORD-02076,2022-07-05,CUST05587,Priya Reddy,29.0,Other,West,Ahmedabad,Furniture,Bed Frame,...,22071.92,0.17,262066.11,66644.99,54.47,EMI,3.0,False,Delivered,4.0
4,ORD-00287,2020-05-07,CUST58035,Ritu Sharma,9.0,Male,Central,Nagpur,Groceries,Cooking Oil,...,320.06,0.31,323.95,31.38,48.81,Net Banking,4.0,False,Pending,4.0


## Initial Inspection

Before any cleaning, I inspect the raw data: shape, column names, data types, missing values, and sample rows.

### Observations

- Shape: **[4310 rows, 21 columns]**
- Column count: **21**
- Data types: mix of `object` (text), `int64`/`float64` (numeric), and `object` for the date column (needs conversion)
- Missing values: present in nearly every column — the dataset is designed to be messy
- Duplicate rows: detected separately

In [11]:
df_before = df.copy()

In [12]:
df = df.dropna(subset = ['order_id'])

#### Empty Row Removal
##### 30 rows had order_id (and every other column) as null — these were fully empty records. Removed them via .dropna(subset=['order_id']), keeping only rows with a valid order identifier. Result: 4,310 → 4,280 rows.

In [13]:
df.info

<bound method DataFrame.info of        order_id  order_date customer_id  customer_name   age  gender   region  \
0     ORD-03903  2024-08-15   CUST32603    Aarav Mehta  39.0   Other  Central   
1     ORD-02198  2022-08-16   CUST07881   Sanjay Verma  28.0    Male     West   
2     ORD-01348  2021-08-29   CUST49296   Pooja Sharma  41.0   Other     West   
3     ORD-02076  2022-07-05   CUST05587    Priya Reddy  29.0   Other     West   
4     ORD-00287  2020-05-07   CUST58035    Ritu Sharma   9.0    Male  Central   
...         ...         ...         ...            ...   ...     ...      ...   
4305  ORD-00056  2020-01-21   CUST85136  Sanjay Sharma  45.0       F     West   
4306  ORD-03458  2024-02-08   CUST09746   Karan Pillai  25.0       m    South   
4307  ORD-01769  2022-02-22   CUST64190    Karan Reddy  39.0    Male     East   
4308  ORD-01738  2022-02-11   CUST50460     Amit Mehta  40.0  Female     East   
4309  ORD-03241  2023-11-01   CUST36164     Nikhil Rao  23.0   Other    North

In [14]:
df = df.drop_duplicates()

### Duplicate Removal
##### Removed exact duplicate rows (identical across all 21 columns). These represent data-export errors rather than genuine repeat purchases, since each order should have a unique order_id. Result: 4,280 → 4,200 rows.

In [15]:
df['gender'] = df['gender'].str.lower()
df['gender'] = df['gender'].replace({'m':'male' , 'f' : 'female'})

C:\Users\majdz\AppData\Local\Temp\ipykernel_10920\3614563979.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['gender'] = df['gender'].str.lower()
C:\Users\majdz\AppData\Local\Temp\ipykernel_10920\3614563979.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['gender'] = df['gender'].replace({'m':'male' , 'f' : 'female'})


In [16]:
df['gender'] = df['gender'].str.lower()


C:\Users\majdz\AppData\Local\Temp\ipykernel_10920\2954947853.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['gender'] = df['gender'].str.lower()


In [17]:
df = df.copy()

In [18]:
df['gender'] = df['gender'].str.lower()
df['gender'] = df['gender'].replace({'m':'male' , 'f': 'female'})

In [19]:
df['gender'].value_counts()

gender
other     1419
male      1408
female    1373
Name: count, dtype: int64

In [20]:
df['gender'] = df['gender'].str.title()

In [21]:
df['gender'].value_counts()

gender
Other     1419
Male      1408
Female    1373
Name: count, dtype: int64

#### The gender column contained 11 distinct labels for only 3 real categories — a mix of case variants (Male, MALE, male) and single-letter abbreviations (M, m, F, f). I first normalized all values to lowercase using .str.lower(), then mapped the abbreviations m → male and f → female using .replace(). Finally, I applied title casing. Result: exactly 3 clean categories — Male, Female, Other.

In [22]:
df.shape

(4200, 21)

In [23]:
print(df['gender'].isnull().sum())
print(df['gender'].value_counts().sum())

0
4200


In [24]:
print("Current shape:", df.shape)
print("Duplicates:", df.duplicated().sum())
print("Null order_id:", df['order_id'].isnull().sum())
print(df['gender'].unique())

Current shape: (4200, 21)
Duplicates: 0
Null order_id: 0
['Other' 'Male' 'Female']


In [25]:
df['order_status'] = df['order_status'].str.title()

### Order Status Standardization
##### order_status appeared with inconsistent casing (e.g., Delivered vs delivered). Applied title-casing to reduce from 9 variants to exactly 5 statuses: Delivered, Shipped, Cancelled, Pending, Returned.

In [26]:
df.shape

(4200, 21)

In [27]:
df['order_date'] = pd.to_datetime(df['order_date'], format = 'mixed', dayfirst=True)

### Date Parsing
##### order_date was stored as text with mixed formats (YYYY-MM-DD and DD/MM/YYYY). Converted to datetime64 using format='mixed' and dayfirst=True (Indian date convention for slash-dates). No NaT values produced.

In [28]:
df.head()

,order_id,order_date,customer_id,customer_name,age,gender,region,city,product_category,product_name,...,unit_price,discount_pct,sales_amount,profit,shipping_cost,payment_method,customer_satisfaction,return_flag,order_status,days_to_ship
0,ORD-03903,2024-08-15,CUST32603,Aarav Mehta,39.0,Other,Central,Raipur,Groceries,Sugar,...,346.32,NaN,4043.14,289.18,56.04,Debit Card,NaN,True,Returned,7.0
1,ORD-02198,2022-08-16,CUST07881,Sanjay Verma,28.0,Male,West,Mumbai,Beauty,Sunscreen,...,858.66,0.08,9147.31,5355.56,64.49,Debit Card,5.0,False,Delivered,2.0
2,ORD-01348,2021-08-29,CUST49296,Pooja Sharma,41.0,Other,West,Surat,Furniture,Sofa,...,25941.32,0.14,110156.97,36956.51,67.65,Debit Card,4.0,False,Delivered,5.0
3,ORD-02076,2022-07-05,CUST05587,Priya Reddy,29.0,Other,West,Ahmedabad,Furniture,Bed Frame,...,22071.92,0.17,262066.11,66644.99,54.47,EMI,3.0,False,Delivered,4.0
4,ORD-00287,2020-05-07,CUST58035,Ritu Sharma,9.0,Male,Central,Nagpur,Groceries,Cooking Oil,...,320.06,0.31,323.95,31.38,48.81,Net Banking,4.0,False,Pending,4.0


In [29]:
print((df['age'] < 13).sum())

146


In [30]:
print((df['age'] > 100).sum())

8


In [31]:
print(df[df['age'] < 13][['age', 'customer_name', 'product_category', 'sales_amount', 'order_date']].head(15))

      age  customer_name product_category  sales_amount order_date
4     9.0    Ritu Sharma        Groceries        323.95 2020-05-07
42   -6.0  Nikhil Kapoor        Furniture      53911.88 2020-07-01
89   10.0   Rohan Pillai         Clothing      12757.93 2023-07-04
107  11.0  Nikhil Sharma      Electronics     172382.02 2024-06-15
108  12.0   Sunita Joshi            Books      10167.90 2022-12-14
147  12.0   Nisha Kapoor      Electronics     154630.39 2023-03-27
199  -5.0   Pooja Kapoor           Sports      22974.82 2023-11-26
214  11.0    Meera Patel         Clothing       3196.43 2024-03-11
305   8.0   Arjun Pillai         Clothing      11602.32 2021-07-05
341   5.0  Ananya Pillai      Electronics     131067.26 2024-12-04
350   2.0   Shreya Mehta         Clothing      12279.25 2022-10-18
351  12.0    Aarav Kumar        Groceries       2253.34 2021-05-28
361  12.0    Pooja Joshi           Sports      66340.96 2020-03-28
424  -5.0     Ritu Patel         Clothing      18790.52 2020-0

In [32]:
print(df[df['age'] > 100][['age', 'customer_name', 'product_category', 'sales_amount', 'order_date']])

        age customer_name product_category  sales_amount order_date
44    999.0   Meera Gupta        Furniture      24230.41 2020-12-09
148   999.0   Ananya Nair           Sports      31129.56 2020-01-28
466   999.0   Meera Patel        Groceries       1215.29 2024-10-17
1010  999.0   Vikram Shah           Sports      81878.90 2024-12-05
1411  150.0   Sanjay Iyer        Groceries        181.61 2022-10-17
2265  150.0   Arjun Singh            Books       7046.17 2024-09-27
3865  999.0  Kavya Kapoor      Electronics   17360314.17 2023-01-14
4168  150.0     Meera Rao         Clothing       2308.91 2021-11-18


In [33]:
import numpy as np
df.loc[(df['age']<13) |  (df['age']>100) , 'age'] = np.nan

## Age Outlier Handling

I found 146 rows with `age < 13` and 8 rows with `age > 100`. Investigation showed these were implausible for online retail customers and likely data-entry errors. I set them to NaN so they can be imputed with the median rather than fabricated.

In [34]:
print(df[['age', 'quantity', 'discount_pct', 'days_to_ship', 'customer_satisfaction']].isnull().sum())

age                      279
quantity                 108
discount_pct             135
days_to_ship              97
customer_satisfaction    346
dtype: int64


In [35]:
df['age'] = df['age'].fillna(df['age'].median())
df['quantity'] = df['quantity'].fillna(df['quantity'].median())
df['discount_pct'] = df['discount_pct'].fillna(0)
df['days_to_ship'] = df['days_to_ship'].fillna(df['days_to_ship'].median())

In [36]:
# Negative values — should most of these be zero or positive?
print("Negative quantity:", (df['quantity'] < 0).sum())
print("Negative unit_price:", (df['unit_price'] < 0).sum())
print("Negative sales_amount:", (df['sales_amount'] < 0).sum())
print("Negative shipping_cost:", (df['shipping_cost'] < 0).sum())
print("Negative discount_pct:", (df['discount_pct'] < 0).sum())
print("Negative age:", (df['age'] < 0).sum())

# Out-of-range values
print("\nDiscount > 1:", (df['discount_pct'] > 1).sum())
print("Age > 100:", (df['age'] > 100).sum())
print("Age < 13:", (df['age'] < 13).sum())
print("Zero quantity:", (df['quantity'] == 0).sum())
print("Zero unit_price:", (df['unit_price'] == 0).sum())

# Profit can legitimately be negative (losses) — check the spread
print("\nProfit range:", df['profit'].min(), "to", df['profit'].max())

Negative quantity: 6
Negative unit_price: 0
Negative sales_amount: 0
Negative shipping_cost: 2
Negative discount_pct: 0
Negative age: 0

Discount > 1: 0
Age > 100: 0
Age < 13: 0
Zero quantity: 12
Zero unit_price: 0

Profit range: 1.56 to 2445648.43


In [37]:
df[df['quantity'] < 0][['order_id', 'quantity', 'unit_price', 'sales_amount', 'profit', 'return_flag', 'order_status']]

,order_id,quantity,unit_price,sales_amount,profit,return_flag,order_status
112,ORD-03804,-1.0,84.21,125.78,8.90,False,Delivered
1129,ORD-03770,-1.0,21887.79,27456.59,3466.29,False,Delivered
1472,ORD-03766,-1.0,3154.09,4594.66,1520.00,False,Delivered
1527,ORD-00533,-1.0,774.58,670.73,287.27,False,Delivered
2112,ORD-04108,-1.0,3079.53,2711.38,987.61,False,Cancelled
2670,ORD-02090,-1.0,1491.38,2043.42,838.22,False,Delivered


In [38]:
df['quantity'] = df['quantity'].abs()

In [39]:
df['quantity'].isnull().sum()

np.int64(0)

In [40]:
(df['quantity']<0).sum()

np.int64(0)

In [41]:
df[df['shipping_cost'] < 0][['order_id', 'shipping_cost', 'quantity', 'unit_price', 'sales_amount', 'profit']]

,order_id,shipping_cost,quantity,unit_price,sales_amount,profit
1878,ORD-03203,-3.11,5.0,30014.40,131495.05,42093.32
2012,ORD-00576,-4.29,10.0,2950.98,23921.04,3847.48


In [42]:
df['shipping_cost'] = df['shipping_cost'].abs()

## Negative Value Corrections

I found negative values in two columns:
- `quantity`: 6 rows with -1 (impossible to order negative items)
- `shipping_cost`: 2 rows with small negatives (e.g., -3.11)

Investigation confirmed these were sign errors (all other columns in the row were positive). I applied `.abs()` to flip the signs without altering the magnitudes.

In [43]:
df[df['quantity'] == 0][['order_id', 'quantity', 'unit_price', 'sales_amount', 'profit', 'order_status']]

,order_id,quantity,unit_price,sales_amount,profit,order_status
179,ORD-00542,0.0,9189.44,8306.29,916.95,Delivered
295,ORD-00034,0.0,315.85,203.41,13.82,Delivered
403,ORD-02274,0.0,803.12,911.76,328.33,Delivered
914,ORD-00414,0.0,1076.68,1005.23,335.77,Cancelled
1081,ORD-03344,0.0,935.36,727.57,161.08,Returned
1495,ORD-01154,0.0,2696.46,2622.68,977.15,Delivered
1914,ORD-03876,0.0,60299.29,91325.87,12487.67,Delivered
2172,ORD-02586,0.0,633.38,637.16,306.98,Delivered
3125,ORD-04072,0.0,16894.52,14763.33,5236.34,Cancelled
3214,ORD-01717,0.0,1351.01,1116.21,341.54,Returned


In [44]:
df = df[df['quantity'] != 0].copy()

In [45]:
print("=== Top 10 by profit ===")
print(df.nlargest(10, 'profit')[['order_id', 'quantity', 'unit_price', 'sales_amount', 'profit', 'product_category', 'order_date']])
print("\n=== Bottom 5 by profit ===")
print(df.nsmallest(5, 'profit')[['order_id', 'quantity', 'unit_price', 'sales_amount', 'profit', 'product_category']])

=== Top 10 by profit ===
       order_id  quantity  unit_price  sales_amount      profit  \
3865  ORD-02560     999.0    18619.45   17360314.17  2445648.43   
1138  ORD-01025     999.0     9380.72    7936261.20  1739854.00   
2582  ORD-00809     999.0     3280.13    2586345.13   921794.60   
1112  ORD-00539     999.0     2567.40    2281458.22   899209.60   
1012  ORD-01242     999.0      545.25     729651.06   258878.20   
2379  ORD-03003      10.0    76450.07    1182709.24   198109.98   
2178  ORD-01732     999.0      709.64     533420.31   141347.64   
3667  ORD-03792      10.0    63505.68    1033433.55   141319.32   
1318  ORD-03751       8.0    67283.39     832824.46   136279.91   
549   ORD-02846      10.0    31142.23     475250.99   133297.06   

     product_category order_date  
3865      Electronics 2023-01-14  
1138           Sports 2021-04-09  
2582         Clothing 2021-01-05  
1112         Clothing 2020-08-25  
1012            Books 2021-07-14  
2379      Electronics 2023-

In [46]:
df['_margin'] = df['profit'] / df['sales_amount']
print("Margin > 1 (impossible):", (df['_margin'] > 1).sum())
print("Margin > 0.5 (unusual):", (df['_margin'] > 0.5).sum())
print("\nTop margins:")
print(df.nlargest(10, '_margin')[['order_id', 'quantity', 'sales_amount', 'profit', '_margin']])
df = df.drop(columns=['_margin'])

Margin > 1 (impossible): 0
Margin > 0.5 (unusual): 233

Top margins:
       order_id  quantity  sales_amount   profit   _margin
4234  ORD-02990       2.0        992.13   703.91  0.709494
2372  ORD-03193       7.0      13177.14  8944.51  0.678790
3203  ORD-03608       8.0       6687.01  4477.03  0.669511
2050  ORD-01332       9.0       2124.79  1395.29  0.656672
439   ORD-00232       4.0        902.31   591.14  0.655141
1429  ORD-03631       6.0      11838.84  7704.41  0.650774
3789  ORD-01205       3.0       1131.28   733.38  0.648275
2144  ORD-00607       8.0      13305.74  8573.91  0.644377
1962  ORD-00340       9.0       4995.33  3217.96  0.644194
2518  ORD-02827       3.0       3386.22  2179.66  0.643685


In [47]:
print("Rows with quantity = 999:", (df['quantity'] == 999).sum())
print("\n=== quantity = 999 rows ===")
print(df[df['quantity'] == 999][['order_id', 'quantity', 'unit_price', 'sales_amount', 'profit', 'product_category', 'order_date']])
print("\nMax non-999 quantity:", df[df['quantity'] != 999]['quantity'].max())

Rows with quantity = 999: 8

=== quantity = 999 rows ===
       order_id  quantity  unit_price  sales_amount      profit  \
242   ORD-00751     999.0      403.63     203951.50    80603.27   
1012  ORD-01242     999.0      545.25     729651.06   258878.20   
1112  ORD-00539     999.0     2567.40    2281458.22   899209.60   
1138  ORD-01025     999.0     9380.72    7936261.20  1739854.00   
1700  ORD-00689     999.0      195.29     140042.31     9617.93   
2178  ORD-01732     999.0      709.64     533420.31   141347.64   
2582  ORD-00809     999.0     3280.13    2586345.13   921794.60   
3865  ORD-02560     999.0    18619.45   17360314.17  2445648.43   

     product_category order_date  
242             Books 2020-12-07  
1012            Books 2021-07-14  
1112         Clothing 2020-08-25  
1138           Sports 2021-04-09  
1700        Groceries 2020-11-08  
2178            Books 2022-02-08  
2582         Clothing 2021-01-05  
3865      Electronics 2023-01-14  

Max non-999 quantity: 1

In [48]:
high_margin = df['profit'] / df['sales_amount'] > 0.5
print(df[high_margin]['product_category'].value_counts())

product_category
Beauty      171
Books        52
Clothing     10
Name: count, dtype: int64


In [49]:
df = df[df['quantity'] != 999].copy()

##  Removing Invalid Quantity Rows

Two issues remained:
- **12 rows** with `quantity = 0` but positive `sales_amount` and `profit` — internally inconsistent (you can't sell 0 items and earn revenue)
- **8 rows** with `quantity = 999` — a placeholder value, since the maximum legitimate order quantity is 10

Both cases have no defensible correction (there's no way to infer what the actual quantity should be), so I dropped these rows entirely.

In [50]:
print("=== FINAL STATE ===")
print("Shape:", df.shape)
print("Duplicates:", df.duplicated().sum())
print("Null order_id:", df['order_id'].isnull().sum())
print("Negative quantity:", (df['quantity'] < 0).sum())
print("Zero quantity:", (df['quantity'] == 0).sum())
print("999 quantity:", (df['quantity'] == 999).sum())
print("Negative shipping_cost:", (df['shipping_cost'] < 0).sum())
print("Negative unit_price:", (df['unit_price'] < 0).sum())
print("Negative sales_amount:", (df['sales_amount'] < 0).sum())

print("\n=== REMAINING NULLS ===")
nulls = df.isnull().sum()
print(nulls[nulls > 0])

print("\n=== DTYPES ===")
print(df.dtypes)

=== FINAL STATE ===
Shape: (4180, 21)
Duplicates: 0
Null order_id: 0
Negative quantity: 0
Zero quantity: 0
999 quantity: 0
Negative shipping_cost: 0
Negative unit_price: 0
Negative sales_amount: 0

=== REMAINING NULLS ===
customer_satisfaction    344
dtype: int64

=== DTYPES ===
order_id                         object
order_date               datetime64[ns]
customer_id                      object
customer_name                    object
age                             float64
gender                           object
region                           object
city                             object
product_category                 object
product_name                     object
quantity                        float64
unit_price                      float64
discount_pct                    float64
sales_amount                    float64
profit                          float64
shipping_cost                   float64
payment_method                   object
customer_satisfaction           float64


In [51]:
df.to_csv(r'C:\internship_oasis_online\OIBSIP\DataAnalytics-L1-EDARetailSales\data\cleaned_retail_sales_dataset.csv' , index = False )